source/dynamical_systems/src/Ring.cpp

```diff
@@ -20,48 +20,92 @@ Ring::Ring() :
+  if (center.get_reference_frame() != this->get_base_frame().get_name()) {
+    if (center.get_reference_frame() != this->get_base_frame().get_reference_frame()) {
+      throw state_representation::exceptions::IncompatibleReferenceFramesException(
+          "The reference frame of the center " + center.get_name() + " in frame " + center.get_reference_frame()
+              + " is incompatible with the base frame of the dynamical system " + this->get_base_frame().get_name()
+              + " in frame " + this->get_base_frame().get_reference_frame() + "."
+      );
+    }
+    this->center_->set_value(this->get_base_frame().inverse() * center);
+  } else {
+    this->center_->set_value(center);
+  }
+  this->set_rotation_offset(this->get_rotation_offset());
 }
 
-CartesianState Ring::compute_dynamics(const CartesianState& state) const {
-  if (this->get_center().is_empty()) {
-    throw exceptions::EmptyAttractorException("The center of the dynamical system is empty.");
-  }
-  // put the point in the reference of the center
-  CartesianPose pose(state);
-  pose = this->get_center().inverse() * pose;
-  // apply the rotation offset
-  pose.set_orientation(pose.get_orientation() * this->get_rotation_offset().conjugate());
+void Ring::set_rotation_offset(const Eigen::Quaterniond& rotation) {
+  auto pose = CartesianPose("rotation", rotation, this->center_->get_value().get_name());
+  this->rotation_offset_->set_value(pose);
+}
 
-  CartesianTwist twist(pose.get_name(), pose.get_reference_frame());
-  double local_field_strength;
-  twist.set_linear_velocity(this->calculate_local_linear_velocity(pose, local_field_strength));
-  twist.set_angular_velocity(this->calculate_local_angular_velocity(pose,
-                                                                    twist.get_linear_velocity(),
-                                                                    local_field_strength));
+Eigen::Quaterniond Ring::get_rotation_offset() const {
+  return this->rotation_offset_->get_value().get_orientation();
+}
 
-  // transform the twist back to the base reference frame
-  return CartesianState(this->get_center()) * twist;
+void Ring::set_base_frame(const state_representation::CartesianState& base_frame) {
+  if (base_frame.is_empty()) {
+    throw state_representation::exceptions::EmptyStateException(base_frame.get_name() + " state is empty");
+  }
+  IDynamicalSystem<state_representation::CartesianState>::set_base_frame(base_frame);
+  // update reference frame of center
+  if (!this->center_->get_value().is_empty()) {
+    auto center = this->center_->get_value();
+    center.set_reference_frame(base_frame.get_name());
+    this->center_->set_value(center);
+  }
+}
+
+void Ring::validate_and_set_parameter(const std::shared_ptr<state_representation::ParameterInterface>& parameter) {
+  if (parameter->get_name() == "center") {
+    this->assert_parameter_valid("center", parameter->get_type());
```

Comment:
seeing this method called so many times, it comes to mind it could be made simpler - you could just have a signature `assert_parameter_valid(std::shared_ptr<state_representation::ParameterInterface>& parameter)`, and then internally do `parameter->get_name()` and `parameter->get_type()`. That would reduce a bit the code duplication and possibility of making a mistake (you don't have to manually write `"center"` twice, since you already confirmed what the parameter name is in the `if` statement)

Reply:
whatever you suggest, as long as i can code a few lines now :star_struck: now but this is a good suggestion obviously